Performance improvements

seancribbs opened the pull request:
Results of a long back-and-forth session with @Vagabond. Taken with basho/erlang_protobuffs#38, this should greatly improve the performance of streaming requests especially.
1. Services are small in number, use `orddict` instead of `dict` to store the state of individual services within a server process.
2. Reduce number of calls to `orddict:fetch` and `orddict:store`. This is accomplished by avoiding storing the service state when it is unchanged, and by keeping the state of a service in streaming mode in the current request state and only storing it when the stream finishes.
3. Reduce number of calls to `gen_tcp:send/2` by buffering outgoing messages up to ~1KB. When the server has no more Erlang messages in its message queue, or the buffer fills, the buffer will be flushed to the socket.

Commit:
Services are small in number, use orddict instead of dict inside riak_api_pb_server.

Commit:
Don't update the service state if unchanged.

Commit:
Memoize the state of the streaming service into the request marker.

Commit:
Add framing module.

Commit:
Buffer outgoing messages so as to call gen_tcp:send/2 fewer times.

Commit:
Fix some dialyzer warnings.

beerriot commented:
The Java client tests fail for me on this PR. Some tests complain about unknown or bad message codes. See here: https://gist.github.com/7b18f6f790a460269a36

seancribbs commented:
@beerriot Was able to reproduce, willfix.

Commit:
Switch to buffering and decoding incoming messages ourselves.

client_ruby_verify and client_python_verify pass on this commit but
client_java_verify hangs and subsequently fails.

Commit:
Drain the incoming buffer in case there are > 1 message in it.

beerriot commented:
Tests are passing for me now, so other than the minor note about `#state.inbuffer`, +1!

Commit:
Cleanup usage/setting of inbuffer, thanks @Vagabond @beerriot.
